File: bank-account/typescript/src/Account.ts
import type { Clock } from './Clock.js';
import { Money } from './Money.js';
import { Transaction } from './Transaction.js';

function formatDate(d: Date): string {
  const y = d.getUTCFullYear();
  const m = String(d.getUTCMonth() + 1).padStart(2, '0');
  const day = String(d.getUTCDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

function formatAmount(amount: number): string {
  return amount.toFixed(2).padStart(7, ' ');
}

export class Account {
  private _balance: Money = Money.zero;
  private _transactions: Transaction[] = [];

  constructor(private readonly clock: Clock) {}

  get balance(): Money { return this._balance; }
  get transactions(): readonly Transaction[] { return this._transactions; }

  deposit(amount: Money): boolean {
    if (!amount.isPositive) return false;
    this._balance = this._balance.plus(amount);
    this._transactions.push(new Transaction(this.clock.today(), amount, this._balance));
    return true;
  }

  withdraw(amount: Money): boolean {
    if (!amount.isPositive) return false;
    if (amount.greaterThan(this._balance)) return false;
    this._balance = this._balance.minus(amount);
    this._transactions.push(
      new Transaction(this.clock.today(), new Money(-amount.amount), this._balance),
    );
    return true;
  }

  printStatement(): string {
    const header = 'Date       | Amount  | Balance';
    const rows = this._transactions.map((t) =>
      `${formatDate(t.date)} | ${formatAmount(t.amount.amount)} | ${formatAmount(t.balanceAfter.amount)}`,
    );
    return [header, ...rows].join('\n');
  }
}

File: bank-account/python/src/bank_account/account.py
from __future__ import annotations

from .clock import Clock
from .money import Money
from .transaction import Transaction


class Account:
    def __init__(self, clock: Clock) -> None:
        self._clock = clock
        self._balance = Money.zero()
        self._transactions: list[Transaction] = []

    @property
    def balance(self) -> Money:
        return self._balance

    @property
    def transactions(self) -> list[Transaction]:
        return list(self._transactions)

    def deposit(self, amount: Money) -> bool:
        if not amount.is_positive:
            return False
        self._balance = self._balance + amount
        self._transactions.append(
            Transaction(self._clock.today(), amount, self._balance)
        )
        return True

    def withdraw(self, amount: Money) -> bool:
        if not amount.is_positive:
            return False
        if amount > self._balance:
            return False
        self._balance = self._balance - amount
        self._transactions.append(
            Transaction(self._clock.today(), -amount, self._balance)
        )
        return True

    def print_statement(self) -> str:
        header = "Date       | Amount  | Balance"
        lines = [header]
        for t in self._transactions:
            date_str = t.date.isoformat()
            amount_str = f"{t.amount.amount:.2f}".rjust(7)
            balance_str = f"{t.balance_after.amount:.2f}".rjust(7)
            lines.append(f"{date_str} | {amount_str} | {balance_str}")
        return "\n".join(lines)
